### PR TITLE
Add unit test for scheduling with full Container

### PR DESCRIPTION
**Pull request by tao-stones**
#### Problem

Discovered during benching prio-graph scheduler, found it mistakenly drops lowest prio tx during scheduling, when Container is full. This seems to be a rare corner case, only happen when Container is full and `schedule()` call wants to put unscheduled IDs back to priority queue. 

Scenario is described in two added unit tests.

#### Summary of Changes
- Added Unit tests to cover the corner case.

Fixes #

### CODE REVIEW

File: core/src/banking_stage/transaction_scheduler/transaction_state_container.rs [hunk 1]
```diff
@@ -89,6 +89,10 @@ pub(crate) trait StateContainer<Tx: TransactionWithMeta> {
     /// Returns `true` if a packet was dropped due to capacity limits.
     fn push_id_into_queue(&mut self, priority_id: TransactionPriorityId) -> bool;
 
+    /// A transaction ID was retrieved for scheduling, when it's not scheduled, put it back into
+    /// priority queue. "put back" shall never need to push out other tx (os is it?)
+    fn put_id_back_into_queue(&mut self, priority_id: TransactionPriorityId);
+
```
Comment (tao-stones) [inline core/src/banking_stage/transaction_scheduler/transaction_state_container.rs]: On the second thought, not too sure if this is the right solution. Is Schedule *putting back* a ID really so different from receiver insert one? Or maybe need a 3rd state that if a transaction is retrieved for scheduling, it's space in Container shall be "reserved", making it not available for new received transactions, until its either scheduled or put-back.

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs [hunk 2]
```diff
@@ -304,7 +304,7 @@ impl<Tx: TransactionWithMeta> Scheduler<Tx> for PrioGraphScheduler<Tx> {
 
         // Push remaining transactions back into the container
         while let Some((id, _)) = self.prio_graph.pop_and_unblock() {
-            container.push_id_into_queue(id);
+            container.put_id_back_into_queue(id);
```
Comment (tao-stones) [inline core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: A separate note: during debugging, I noticed those IDs being put back are earliest txs, eg `{priority: 1; id: 0}`, `{priority: 1; id: 1}`... I would think given same priority, txs received earlier should be scheduled earlier. This doesn't seem the case tho.

Comment (apfitzge) [reply core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: ids are recycled in a stack because that's how `Slab` finds them; it's an index

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs [hunk 3]
```diff
@@ -931,4 +931,107 @@ mod tests {
         assert_eq!(scheduling_summary.num_unschedulable, 0);
         assert_eq!(collect_work(&work_receivers[0]).1, vec![vec![2], vec![0]]);
     }
+
+    fn create_container_with_capacity(
```
Comment (apfitzge) [inline core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: can we not share some impl with `create_container` above?

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs [hunk 4]
```diff
@@ -931,4 +931,107 @@ mod tests {
...
+                packet,
+                compute_unit_price,
+                TEST_TRANSACTION_COST,
+            );
+        }
+
+        container
+    }
+
+    fn test_schedulling(
```
Comment (apfitzge) [inline core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: can name be more descriptive? there are other scheduling tests above

**Comment by apfitzge** [comment]
is the title of the PR still true after the rebase? if not, let's edit it

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs (hunk 3 above)
Comment (tao-stones) [reply core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: Yes. Duplicating code to make this commit easy to read for its intention. Will refactor this function and fold the new "_small" and _large" tests into one.

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs (hunk 2 above)
Comment (tao-stones) [reply core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs]: I meant the `ids` that are `pop_and_unblock()` from prio_graph, not before being reinserted into Container.

**Review by apfitzge — APPROVED** [review]
